### ISSUE TITLE
Serializing to JSON

### ISSUE THREAD

**Issue by tarelli** [openworm/org.geppetto.python#4]
Hi @aranega I am trying to serialize to JSON a model built programmatically
```
    flib = pygeppetto.GeppettoLibrary(name='netpynelib')
    # We create a GeppettoModel instance and we set a name a assign a lib
    root = pygeppetto.GeppettoModel(name='NetPyNEModel', libraries=[flib])
    rset = ResourceSet()  # We have a resource set
    uri = URI('netpyne_model')
    rset.resource_factory['json'] = lambda uri: JsonResource(uri)  # we register the factory for '.json' extensions
    resource = rset.create_resource(uri)
    resource.append(root)
    resource.save()
```
but what I get is XMI
```
<?xml version='1.0' encoding='UTF-8'?>
<gep:GeppettoModel xmlns:gep="https://raw.githubusercontent.com/openworm/org.geppetto.model/development/src/main/resources/geppettoModel.ecore" xmlns:xmi="http://www.omg.org/XMI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmi:version="2.0" name="NetPyNEModel">
  <libraries xsi:type="gep:GeppettoLibrary" name="netpynelib"/>
</gep:GeppettoModel>
```

What am I doing wrong? 
Also what would be the best way to output the result of the serialisation to string? In Java I do something like this:

```
		StringWriter sw = new StringWriter();
		WriteableOutputStream outputStream = new WriteableOutputStream(sw, "UTF-8");
```

Thanks!

**Comment by aranega** [comment]
Hi @tarelli,

Sorry, for the delay, I wanted to check some points first, and I found some issues with the JSON deserialization (serialization was fine) and the way the inner streams are used for URI. Upgrade your `pyecore-py2` version first, the last fresh one is the 0.7.3. 

Regarding your first point, written like this, the `JsonResource` is registered for `.json` extensions only, that's why your uri: `netpyne_model` is not identified as a json file. To enables the resource set selecting a `JsonResource`, you just need to rename your uri from `netpyne_model` to `netpyne_model.json`. By default, the `XMIResource` is registered for all (or no) extensions, that's why an XMI resource was used. You can override this by overriding the `*` extension in the `resource_factory`:

```python
rset.resource_factory['*'] = lambda uri: JsonResource(uri, indent=2)
``` 
Also, you can use `indent` to precise the space that should be used for the JSON resource formatting. By default I put `0` as the JSON serialization is more verbose than XMI (some namespace are repeated many times through the file while in XMI, they are 'factorized' in the file "header").

If you want to output the result of the serialization to a string, you need to write your own `URI`. I put an example in PyEcore ["documentation"](https://github.com/pyecore/pyecore/#creating-your-own-uri) with the new version. To deal with this, an implementation could be the following:

```python
class StringURI(URI):
    def __init__(self, uri, text=None):
        super(StringURI, self).__init__(uri)
        if text is not None:
            self.__stream = StringIO(text)

    def getvalue(self):
        return self.__stream.getvalue()

    def create_instream(self):
        return self.__stream

    def create_outstream(self):
        self.__stream = StringIO()
        return self.__stream
```

Then, you can use `StringURI` instead of `URI`, like this for serialization:

```python
uri = StringURI('netpyne_model.json')  # an URI id is still required to create the appropriate Resource
resource = rset.create_resource(uri)
resource.append(root)
resource.save()
print(uri.getvalue())
```

and for deserialization

```python
uri = StringURI('my_uri.json', text=myStrModel)  # an URI id is still required to create the appropriate Resource
resource = rset.get_resource(uri)
model_root = resource.contents[0]

# or if you prefer
uri = StringURI('my_uri.json', text=myStrModel)
resource = rset.create_resource(uri)
resource.load()
model_root = resource.contents[0]
```

PS: I also had to modify a little bit the pygeppetto generated code, you can pull the new version from `developement`.

**Comment by tarelli** [comment]
Thanks so much @aranega!
To make it work I had to modify it like this:

```python
class StringURI(URI):

    def __init__(self, uri, text=None):
        super(StringURI, self).__init__(uri)
        if text is not None:
            self.__stream = io.StringIO(text)

    def getvalue(self):
        return self.__stream.getvalue()

    def create_instream(self):
        self.__stream = open(self.plain, 'rb')
        return self.__stream

    def create_outstream(self):
        self.__stream = StringIO()
        return self.__stream
```
What I get now is the following:

```json
{
  "libraries": [
    {
      "eClass": "https://raw.githubusercontent.com/openworm/org.geppetto.model/development/src/main/resources/geppettoModel.ecore#//GeppettoLibrary", 
      "name": "netpynelib"
    }
  ], 
  "eClass": "https://raw.githubusercontent.com/openworm/org.geppetto.model/development/src/main/resources/geppettoModel.ecore#//GeppettoModel", 
  "name": "NetPyNEModel"
}
```
This is great because it's very close to what we get when we use the Java server. The only difference is that in Java we were able to introduce a custom serialiser to make the output representation more compact:

```java
	private static class GeppettoCustomTypeSerializer implements TypeSerializer
	{

		@Override
		public void serialize(EClass eClass, JsonGenerator jg, SerializerProvider provider) throws IOException
		{

			jg.writeStringField("eClass", eClass.getName());
		}

	}
```
which essentially produces `"eClass" : "GeppettModel"` instead of having the whole URL. Is it possible to do something analogue with pyecore?
Thanks again so much, this is a great development for us!

**Comment by aranega** [comment]
@tarelli Aaah, you read from a file and you write to a string. Sorry, I missed that.

Currently, it is not possible to add a dedicated serializer for the JSON resource, I can see if I can add such a mechanism for the JSON serializer or a way of properly monkey-patch the `JsonResource`. Could you send me a full JSON example?
I'm surprised that using only the `EClass` name works for reloading the JSON after a serialization. Some information must be set in another place (perhaps the root) otherwise, there is no information about the handled metamodel, and loading a model from an existing JSON seems difficult. This is even more surprising as the Geppetto metamodel is split in several sub-packages.

**Comment by aranega** [comment]
@tarelli I got it, I searched about the Java JSON serializer (I assumed `emfjson-jackson`), and I think I found the information I needed. So, correct me if I'm wrong, but if implemented your own custom serializer, you probably also have your  own deserializer that exclusively uses the Geppetto metamodel? (like the example in this [doc](http://emfjson.org/projects/jackson/latest/))

In this spirit, I modified, in the `develop` branch of PyEcore only at the moment, the `JsonResource` so the two methods responsible for the eclass reference serialization and deserialization can be overriden.

This allowed me to create a dedicated `Resource` like this:

```python
from backports.functools_lru_cache import lru_cache
from chainmap import ChainMap
from pyecore.resources.json import JsonResource
import model as pygeppetto

class GeppettoResource(JsonResource):
    packages = [pygeppetto.eClassifiers,
                pygeppetto.datasources.eClassifiers,
                pygeppetto.types.eClassifiers,
                pygeppetto.values.eClassifiers,
                pygeppetto.variables.eClassifiers]
    chain = ChainMap(*packages)

    def serialize_eclass(self, eclass):
        return eclass.name

    @lru_cache()
    def resolve_eclass(self, uri):
        return self.chain.get(uri)
```
* `resolve_eclass` is responsible for the `EClass` resolution regarding the eclass uri from the resource to deserialize
* ` serialize_eclass` is responsible for the `EClass` reference serialization

The `resolve_eclass` method searches through all the `eClassifiers` from the main package and its subpackages for an `EClass` regarding only its name. The `lru_cache` use is not mandatory, but in some case, it can really improve performances (not sure in your case). The `serialize_eclass` method only returns the `EClass` name, as does your dedicated Java serializer.

You can register this new resource as any other:

```python
# Here, the GeppettoResource is registered for '.geppetto' extensions only
rset.resource_factory['geppetto'] = lambda uri: GeppettoResource(uri, indent=2)
```
Is the production of a new dedicated Resource could be acceptable for you?  I prefered this way as it still allows you to handle other kind of JSON resources with the 'original' behavior.

I'm still conducting some experimentation to ensure everything's fine with the Java JSON serializer (I have doubts about `EEnums`), and if everything runs smoothly, I will release a new PyEcore version today.

**Comment by tarelli** [comment]
@aranega you are amazing. Thank you, that is very acceptable, I will test the new release when it comes out!

**Comment by aranega** [comment]
@tarelli I just released the new version: 0.7.4. In the same time, I fixed some challenges with `EEnums`. I tried to import JSON files obtained from XMI translated to JSON using `emfjson-jackson`, everything's fine. Please not that it may still have some issues as it is difficult to be certain that all the cases are covered, I will try to deal with them as quickly as possible.

EDIT> I forgot, but I'm in vacation until the end of the next week, and I was told to let my laptop at home. I will be able to answer, but not to check/code stuffs as quickly as I could.

**Comment by tarelli** [comment]
That worked!
